File: context_window_economics/store.py
"""Append-only JSONL store for metering records and settlements.

Provides persistent storage for CMRs, settlement receipts, and deposit records.
All writes are append-only for auditability.
"""
import json
import os
from pathlib import Path
from typing import Any, Dict, List, Optional, Type, TypeVar

from .schema import CostMeteringRecord, DepositRecord, SettlementProposal

T = TypeVar("T")

DEFAULT_STORE_DIR = ".cwep"
CMR_FILE = "metering.jsonl"
SETTLEMENT_FILE = "settlements.jsonl"
DEPOSIT_FILE = "deposits.jsonl"


class CWEPStore:
    """Append-only JSONL store for CWEP records.

    Files are stored in a directory (default: .cwep/ in working dir).
    Each record type gets its own JSONL file.
    """

    def __init__(self, store_dir: str = DEFAULT_STORE_DIR):
        self.store_dir = Path(store_dir)
        self.store_dir.mkdir(parents=True, exist_ok=True)
        self._cmr_path = self.store_dir / CMR_FILE
        self._settlement_path = self.store_dir / SETTLEMENT_FILE
        self._deposit_path = self.store_dir / DEPOSIT_FILE

    # ----- CMR operations -----

    def append_cmr(self, cmr: CostMeteringRecord) -> None:
        """Append a CMR to the metering log."""
        self._append(self._cmr_path, cmr.to_dict())

    def read_cmrs(self, limit: int = 0) -> List[CostMeteringRecord]:
        """Read CMRs from the store.

        Args:
            limit: Maximum number to return (0 = all, most recent first).
        """
        records = self._read_all(self._cmr_path)
        if limit > 0:
            records = records[-limit:]
        return [CostMeteringRecord.from_dict(r) for r in records]

    def cmr_count(self) -> int:
        """Count CMRs in the store."""
        return self._count_lines(self._cmr_path)

    # ----- Settlement operations -----

    def append_settlement(self, data: Dict[str, Any]) -> None:
        """Append a settlement record."""
        self._append(self._settlement_path, data)

    def read_settlements(self, limit: int = 0) -> List[Dict[str, Any]]:
        """Read settlement records."""
        records = self._read_all(self._settlement_path)
        if limit > 0:
            records = records[-limit:]
        return records

    def settlement_count(self) -> int:
        return self._count_lines(self._settlement_path)

    # ----- Deposit operations -----

    def append_deposit(self, deposit: DepositRecord) -> None:
        """Append a deposit record."""
        self._append(self._deposit_path, deposit.to_dict())

    def read_deposits(self, limit: int = 0) -> List[DepositRecord]:
        """Read deposit records."""
        records = self._read_all(self._deposit_path)
        if limit > 0:
            records = records[-limit:]
        return [DepositRecord.from_dict(r) for r in records]

    def deposit_count(self) -> int:
        return self._count_lines(self._deposit_path)

    # ----- Aggregate statistics -----

    def statistics(self) -> Dict[str, Any]:
        """Compute aggregate statistics across all stored records."""
        cmrs = self.read_cmrs()
        total_cost = sum(c.totals.total_cost_usd for c in cmrs)
        total_tokens = sum(c.totals.total_tokens for c in cmrs)

        # Per-agent cost breakdown
        agent_costs: Dict[str, float] = {}
        for c in cmrs:
            if c.requestor:
                aid = c.requestor.agent_id
                agent_costs[aid] = agent_costs.get(aid, 0.0) + c.totals.requestor_incurred_usd
            if c.responder:
                aid = c.responder.agent_id
                agent_costs[aid] = agent_costs.get(aid, 0.0) + c.totals.responder_incurred_usd

        return {
            "cmr_count": len(cmrs),
            "settlement_count": self.settlement_count(),
            "deposit_count": self.deposit_count(),
            "total_cost_usd": total_cost,
            "total_tokens": total_tokens,
            "agent_costs": agent_costs,
            "store_dir": str(self.store_dir),
        }

    # ----- Internal -----

    def _append(self, path: Path, data: Dict[str, Any]) -> None:
        """Append a JSON record as a single line."""
        with open(path, "a", encoding="utf-8") as f:
            f.write(json.dumps(data, default=str) + "\n")

    def _read_all(self, path: Path) -> List[Dict[str, Any]]:
        """Read all JSONL records from a file."""
        if not path.exists():
            return []
        records = []
        with open(path, "r", encoding="utf-8") as f:
            for line in f:
                line = line.strip()
                if line:
                    records.append(json.loads(line))
        return records

    def _count_lines(self, path: Path) -> int:
        """Count non-empty lines in a JSONL file."""
        if not path.exists():
            return 0
        count = 0
        with open(path, "r", encoding="utf-8") as f:
            for line in f:
                if line.strip():
                    count += 1
        return count

File: typescript/src/store.ts
import {
  appendFileSync,
  existsSync,
  mkdirSync,
  readFileSync,
  statSync,
} from "node:fs";
import { dirname, join } from "node:path";
import {
  CostMeteringRecord,
  type CostMeteringRecordData,
  DepositRecord,
  type DepositRecordData,
} from "./types";

const DEFAULT_STORE_DIR = ".cwep";
const CMR_FILE = "metering.jsonl";
const SETTLEMENT_FILE = "settlements.jsonl";
const DEPOSIT_FILE = "deposits.jsonl";

export class CWEPStore {
  readonly storeDir: string;
  private readonly _cmrPath: string;
  private readonly _settlementPath: string;
  private readonly _depositPath: string;
  private _lastParseErrors: number = 0;

  constructor(storeDir: string = DEFAULT_STORE_DIR) {
    this.storeDir = storeDir;
    mkdirSync(storeDir, { recursive: true });
    this._cmrPath = join(storeDir, CMR_FILE);
    this._settlementPath = join(storeDir, SETTLEMENT_FILE);
    this._depositPath = join(storeDir, DEPOSIT_FILE);
  }

  get lastParseErrors(): number {
    return this._lastParseErrors;
  }

  // ----- CMR operations -----

  appendCmr(cmr: CostMeteringRecord): void {
    this._append(this._cmrPath, cmr.toDict() as unknown as Record<string, unknown>);
  }

  readCmrs(limit: number = 0): CostMeteringRecord[] {
    let records = this._readAll(this._cmrPath);
    if (limit > 0) {
      records = records.slice(-limit);
    }
    return records.map((r) =>
      CostMeteringRecord.fromDict(r as unknown as CostMeteringRecordData)
    );
  }

  cmrCount(): number {
    return this._countLines(this._cmrPath);
  }

  // ----- Settlement operations -----

  appendSettlement(data: Record<string, unknown>): void {
    this._append(this._settlementPath, data);
  }

  readSettlements(limit: number = 0): Record<string, unknown>[] {
    let records = this._readAll(this._settlementPath);
    if (limit > 0) {
      records = records.slice(-limit);
    }
    return records;
  }

  settlementCount(): number {
    return this._countLines(this._settlementPath);
  }

  // ----- Deposit operations -----

  appendDeposit(deposit: DepositRecord): void {
    this._append(this._depositPath, deposit.toDict() as unknown as Record<string, unknown>);
  }

  readDeposits(limit: number = 0): DepositRecord[] {
    let records = this._readAll(this._depositPath);
    if (limit > 0) {
      records = records.slice(-limit);
    }
    return records.map((r) => DepositRecord.fromDict(r as unknown as DepositRecordData));
  }

  depositCount(): number {
    return this._countLines(this._depositPath);
  }

  // ----- Aggregate statistics -----

  statistics(): Record<string, unknown> {
    const cmrs = this.readCmrs();
    let totalCost = 0;
    let totalTokens = 0;
    const agentCosts: Record<string, number> = {};

    for (const c of cmrs) {
      totalCost += c.totals.totalCostUsd;
      totalTokens += c.totals.totalTokens;

      if (c.requestor) {
        const aid = c.requestor.agentId;
        agentCosts[aid] =
          (agentCosts[aid] ?? 0.0) + c.totals.requestorIncurredUsd;
      }
      if (c.responder) {
        const aid = c.responder.agentId;
        agentCosts[aid] =
          (agentCosts[aid] ?? 0.0) + c.totals.responderIncurredUsd;
      }
    }

    return {
      cmr_count: cmrs.length,
      settlement_count: this.settlementCount(),
      deposit_count: this.depositCount(),
      total_cost_usd: totalCost,
      total_tokens: totalTokens,
      agent_costs: agentCosts,
      store_dir: this.storeDir,
    };
  }

  // ----- Internal -----

  private _append(path: string, data: Record<string, unknown>): void {
    const dir = dirname(path);
    if (dir && dir !== ".") {
      mkdirSync(dir, { recursive: true });
    }
    appendFileSync(path, JSON.stringify(data) + "\n", "utf-8");
  }

  private _readAll(path: string): Record<string, unknown>[] {
    if (!existsSync(path)) return [];
    const content = readFileSync(path, "utf-8");
    const records: Record<string, unknown>[] = [];
    let parseErrors = 0;
    for (const line of content.split("\n")) {
      const trimmed = line.trim();
      if (!trimmed) continue;
      try {
        records.push(JSON.parse(trimmed));
      } catch {
        parseErrors++;
      }
    }
    this._lastParseErrors = parseErrors;
    return records;
  }

  private _countLines(path: string): number {
    if (!existsSync(path)) return 0;
    const content = readFileSync(path, "utf-8");
    let count = 0;
    for (const line of content.split("\n")) {
      if (line.trim()) count++;
    }
    return count;
  }
}
